### ISSUE TITLE
hidden threadJoin

### ISSUE THREAD

**Issue by WilkAndy** [idaholab/moose#10438]
## Description

I have no idea if this is a small bug, or just harmless (in which case just close this issue), but i just noticed it in the compilation:
```
moose/framework/include/userobject/GeneralUserObject.h:54:16: warning: âvirtual void GeneralUserObject::threadJoin(const UserObject&)â was hidden [-Woverloaded-virtual]
   virtual void threadJoin(const UserObject &) final;
                ^
In file included from moose/framework/include/vectorpostprocessors/LineFunctionSampler.h:19:0,
                 from moose/projects_andy/moose/framework/src/vectorpostprocessors/LineFunctionSampler.C:15:
moose/framework/include/vectorpostprocessors/SamplerBase.h:99:16: warning:   by âSamplerBase::threadJoin(const SamplerBase&)â [-Woverloaded-virtual]
   virtual void threadJoin(const SamplerBase & y);

```

**Comment by permcody** [comment]
Not a bug, but it's arguably a design flaw. I've seen this before and it should be removed or cleaned up.

**Comment by permcody** [comment]
@WilkAndy, What compiler/machine combination are you using? I can't replicate this. I agree the code is screwy, but I don't see that warning.

**Comment by WilkAndy** [comment]
mpicc -v
Using built-in specs.
COLLECT_GCC=/tools/gcc/4.9.3/bin/gcc
COLLECT_LTO_WRAPPER=/opt/tools/gcc/4.9.3/bin/../libexec/gcc/x86_64-unknown-linux-gnu/4.9.3/lto-wrapper
Target: x86_64-unknown-linux-gnu
Configured with: ../gcc-4.9.3/configure --prefix=/tools/gcc/4.9.3 --with-gmp=/tools/gmp/5.1.1 --with-mpfr=/tools/mpfr/3.1.2 --with-mpc=/tools/mpc/1.0
Thread model: posix
gcc version 4.9.3 (GCC)
wil04q@cemvmc:/apps/geomodel/moose/projects_andy/moose/modules> mpicc --version
gcc (GCC) 4.9.3
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


 lsb_release -a
LSB Version:    core-2.0-noarch:core-3.2-noarch:core-4.0-noarch:core-2.0-x86_64:core-3.2-x86_64:core-4.0-x86_64:desktop-4.0-amd64:desktop-4.0-noarch:graphics-2.0-amd64:graphics-2.0-noarch:graphics-3.2-amd64:graphics-3.2-noarch:graphics-4.0-amd64:graphics-4.0-noarch
Distributor ID: SUSE LINUX
Description:    SUSE Linux Enterprise Server 11 (x86_64)
Release:        11
Codename:       n/a

**Comment by YaqiWang** [comment]
Does adding `override` keyword to that function remove the warning?

**Comment by permcody** [comment]
It's not actually declared in this class. The problem is that it's multiply inheriting two interfaces with the threadJoin methods. I don't think we actually hold references to the object through either of the intermediate classes so the design is probably what's at fault here. It would be handy if I could replicate this error to figure out the best course of action. Clang seems to be happy with the using statement that was added here.

**Comment by permcody** [comment]
Fixed by #10717
